File: 0144-binary-tree-preorder-traversal/0144-binary-tree-preorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution {
public:
    void traversal(TreeNode* root, vector<int>& ans) {
        if (root) {
            ans.push_back(root->val);
            if (root->left)
                traversal(root->left, ans);
            if (root->right)
                traversal(root->right, ans);
        }
        return;
    }
    //iterative approach
    vector<int> preorderTraversal(TreeNode* root) {
        vector<int> ans;
        //if (root == NULL)
          //  return {};
        //traversal(root, ans);
    //Iterative approach
    if(root==NULL)return ans;
    stack<TreeNode*>st;
    st.push(root);
    while(!st.empty() ){
        root=st.top();
        st.pop();
        ans.push_back(root->val);
        if(root->right) st.push(root->right);
        if(root->left) st.push(root->left);
    }
        return ans;
    }
};

File: 0144-binary-tree-preorder-traversal/0144-binary-tree-preorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
//iterative approach
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer>pre=new ArrayList<>();
        if(root==null)return pre;
        Stack<TreeNode>stk=new Stack<>();
        stk.push(root);
        while(!stk.isEmpty()){
            root=stk.pop();
            pre.add(root.val);
            if(root.right!=null)
                stk.push(root.right);
            if(root.left!=null)
                stk.push(root.left); 
        }
        return pre;
    }
}